libcob/jor.c

```diff
@@ -0,0 +1,464 @@
+	*jor_position++ = field;		\
+	*jor_position++ = type
+
+#define OPCODE_NEW_NAME 0
+#define OPCODE_LOCAL_FIELD 1
+
+#define TYPE_RECORD     0
+#define TYPE_UINT8      1
+#define TYPE_INT8       2
+#define TYPE_UINT16     3
+#define TYPE_INT16      4
+#define TYPE_UINT32     5
+#define TYPE_INT32      6
+#define TYPE_UINT64     7
+#define TYPE_INT64      8
+#define TYPE_FLOAT      9
+#define TYPE_STRING8   10
+#define TYPE_STRING16  11
+
+static int jor_field_name (const char* s)
+{
+	int id = jor_name_counter++;
+	int len = strlen (s);
+	RECORD_BEGIN (OPCODE_NEW_NAME);
+	*jor_position++ = id;
+	*jor_position++ = len;
+	memcpy (jor_position, s, len);
+	jor_position += len;
+	RECORD_END ();
+	return id;
+}
+
+static void jor_field_record (int level, int field)
+{
+	RECORD_BEGIN (OPCODE_LOCAL_FIELD);
+	RECORD_FIELD (level, field, TYPE_RECORD);
+	RECORD_END ();
+}
+
+static void jor_field_uint8 (int level, int field, cob_u8_t v)
+{
+	RECORD_BEGIN (OPCODE_LOCAL_FIELD);
+	RECORD_FIELD (level, field, TYPE_UINT8);
+	*jor_position++ = v;
+	RECORD_END ();
+}
+static void jor_field_uint32 (int level, int field, cob_u32_t v)
+{
+	RECORD_BEGIN (OPCODE_LOCAL_FIELD);
+	RECORD_FIELD (level, field, TYPE_UINT32);
+	* ( (cob_u32_t*)jor_position) = v;
+	jor_position += 4;
+	RECORD_END ();
+}
+static void jor_field_string8 (int level, int field, const char* v)
+{
+	int len = v != NULL ? strlen(v) : 0;
+	RECORD_BEGIN (OPCODE_LOCAL_FIELD);
+	RECORD_FIELD (level, field, TYPE_STRING8);
+	*jor_position++ = len;
+	if (len>0){
+		memcpy ( jor_position, v, len);
+		jor_position += len;
+	}
+	RECORD_END ();
+}
+
+static
+char* jor_allocate(int *size)
+{
+	return cob_malloc(*size);
+}
+
+static
+void jor_save (const char* filename, char* buffer, int len)
+{
+	FILE *fd = fopen (filename, "w");
+	const char* s = jor_buffer;
+	fwrite (s, len, 1, fd);
+	fclose (fd);
+}
+
+static struct cob_jor_funcs jor_funcs = {
+	jor_save,
+	jor_allocate
+};
+
+void cob_jor_set_funcs (struct cob_jor_funcs *f)
+{
+	jor_funcs.save = f->save;
+	jor_funcs.allocate = f->allocate;
+}
+
+void cob_init_jor (cob_global *lptr, cob_settings *sptr,
+		   int cob_argc, char** cob_argv)
+{
+	int i;
+
+	if (!!cobsetptr) return ;
+
+	cobglobptr = lptr;
+	cobsetptr  = sptr;
+	cob_jor_filename = cob_strdup (cobsetptr->cob_jor_filename);
+
+	/* Check that these fields have been correctly initialized
+	   by the developer. */
+	if (COB_JOR_FILE_OPERATIONS != COB_JOR_AFTER_LAST_OPERATION){
+		fprintf (stderr, "COB_JOR_FILE_OPERATIONS = %d\n",
+			 COB_JOR_FILE_OPERATIONS);
+		fprintf (stderr, "COB_JOR_AFTER_LAST_OPERATION = %d\n",
+			 COB_JOR_AFTER_LAST_OPERATION);
+		exit (2);
+	}
+	if ( field_file_op_name[COB_JOR_FILE_OPERATIONS-1] == NULL ){
+		fprintf (stderr,
+			 "field_file_op_name[%d] not initialized\n",
+			 COB_JOR_FILE_OPERATIONS-1);
+		exit (2);
+	}
+
+	if (!cobsetptr->cob_jor_enable &&
+	    /* testsuite clears COB_JOR_ENABLE... */
+	    !getenv ("COB_JOR_ENABLED")) return ;
```

Comment:
that will be already be handled in the settings pointer, so this part can be removed